### PR TITLE
[Distributed] Fix distributed training hang with multiple samplers

### PR DESCRIPTION
**Pull request by VoVAllen**
## Description
Not sure about the real root cause. But related to the multiprocessing Manager and Barrier

## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

**Comment by zheng-da** [comment]
does this work?

**Comment by VoVAllen** [comment]
Not really

**Comment by VoVAllen** [comment]
I believe this is a Python bug described in https://stackoverflow.com/a/59662879/12552906

**Comment by VoVAllen** [comment]
Related issue https://stackoverflow.com/questions/58843576/multiprocessing-pool-hangs-indefinitely-after-close-join

**Comment by VoVAllen** [comment]
https://stackoverflow.com/questions/20574810/multiprocessing-reading-big-input-data-program-hangs

File: python/dgl/distributed/dist_context.py [hunk 1]
```diff
@@ -41,6 +45,126 @@ def _init_rpc(ip_config, num_servers, max_queue_size, net_type, role, num_thread
...
+    INIT_RPC = 0  # Not used in the task queue
+    SET_COLLATE_FN = 1
+    CALL_BARRIER = 2
+    DELETE_COLLATE_FN = 3
+    CALL_COLLATE_FN = 4
+    CALL_FN_ALL_WORKERS = 5
+    FINALIZE_POOL = 6
+
+
+def init_process(_, rpc_config, mp_contexts):
```
Comment (zheng-da) [inline python/dgl/distributed/dist_context.py:60]: why do you want to have the first argument while not using it?

Comment (VoVAllen) [reply python/dgl/distributed/dist_context.py]: Removed

**Comment by VoVAllen** [comment]
@zheng-da Fixed CI

File: python/dgl/distributed/dist_dataloader.py [hunk 2]
```diff
@@ -140,41 +90,38 @@ def __init__(self, dataset, batch_size, shuffle=False, collate_fn=None, drop_las
...
-            barrier = m.Barrier(self.num_workers)
-            for _ in range(self.num_workers):
-                results.append(self.pool.apply_async(cleanup_fn, args=(barrier, self.name,)))
-            for res in results:
-                res.get()
+            self.pool.delete_collate_fn(self.name)
 
     def __next__(self):
-        num_reqs = self.queue_size - self.num_pending
+        if self.pool is None:
```
Comment (zheng-da) [inline python/dgl/distributed/dist_dataloader.py:103]: why `self.pool` is None?

File: python/dgl/distributed/dist_dataloader.py [hunk 3]
```diff
@@ -112,16 +66,12 @@ def __init__(self, dataset, batch_size, shuffle=False, collate_fn=None, drop_las
...
         self.batch_size = batch_size
         self.num_pending = 0
         self.collate_fn = collate_fn
         self.current_pos = 0
-        if self.pool is not None:
-            m = mp.Manager()
-            self.queue = m.Queue(maxsize=queue_size)
-        else:
-            self.queue = Queue(maxsize=queue_size)
+        self.queue = []
```
Comment (zheng-da) [inline python/dgl/distributed/dist_dataloader.py:74]: this is used when multiprocessing is disabled? if so, can you comment it?

File: src/rpc/rpc.cc [hunk 4]
```diff
@@ -339,7 +339,6 @@ DGL_REGISTER_GLOBAL("distributed.rpc._CAPI_DGLRPCHandleSignal")
   sigemptyset(&sigHandler.sa_mask);
   sigHandler.sa_flags = 0;
   sigaction(SIGINT, &sigHandler, nullptr);
-  sigaction(SIGTERM, &sigHandler, nullptr);
```
Comment (zheng-da) [inline src/rpc/rpc.cc:342]: why do you delete this?

**Comment by zheng-da** [comment]
overall, looks good to me.

**Review by zheng-da — APPROVED** [review]

File: python/dgl/distributed/dist_dataloader.py (hunk 3 above)
Comment (VoVAllen) [reply python/dgl/distributed/dist_dataloader.py]: Added

File: src/rpc/rpc.cc (hunk 4 above)
Comment (VoVAllen) [reply src/rpc/rpc.cc]: I added it back, for the shared memory cleaning. But I found catching SIGTERM prevents python to show the correct Exceptions sometime.

**Pull request merged**